src/Plutarch/Test/QuickCheck.hs

```diff
@@ -77,120 +84,226 @@ import Test.QuickCheck (
-    fromPFun' _ f (TestableTerm x) = TestableTerm $ f # x
+instance PUnLam' fin fin 'True where
+    pUnLam' x = TestableTerm x
 
--- | @since 2.0.0
-instance
-    forall
-        (a :: S -> Type)
-        (b :: S -> Type)
-        (c :: S -> Type)
-        (d :: S -> Type)
-        (e :: S -> Type).
-    ( c ~ (d :--> e)
-    , PLift a
-    , PLift b
-    , FromPFunN d e (PTT c)
-    ) =>
-    FromPFunN (a :--> b) c 'PFunction
-    where
-    fromPFun' _ f (PFun _ _ (unTestableTerm -> x)) =
-        fromPFun' (Proxy @(PTT c)) $ f # x
+{- | Constraint for `PUnLam'` that will give a better type error message.
 
--- | @since 2.0.0
-instance
-    forall (a :: S -> Type) (b :: S -> Type) (c :: S -> Type) (d :: S -> Type).
-    ( b ~ (c :--> d)
-    , FromPFunN c d (PTT b)
-    ) =>
-    FromPFunN a b 'PTerm
-    where
-    fromPFun' _ f (TestableTerm x) = fromPFun' (Proxy @(PTT b)) $ f # x
+ @since 2.1.0
+-}
+type PUnLam fin p = (PUnLam' fin p (IsFinal fin p), CheckReturn fin p)
+
+{- | Wrapper for @pUnLam'@. Only required the final return type.
+
+ @since 2.1.0
+-}
+punlam' ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam' = pUnLam' @fin @p @(IsFinal fin p)
+
+{- | Wrapper for @pUnLam'@. Same as @punlam'@ but evaluates the given
+     Plutarch function before the conversion. It will throw an error
+     if evaluation failes.
+
+ @since 2.1.0
+-}
+punlam ::
+    forall (fin :: S -> Type) (p :: S -> Type).
+    PUnLam fin p =>
+    (forall s. Term s p) ->
+    PUnLamHask fin p
+punlam pf = punlam' @fin (loudEval pf)
 
-{- | Converts Plutarch Functions into `Testable` Haskell function of
-  TestableTerms.
+{- | Converts Plutarch Functions into Haskell function of
+     TestableTerms. Then, wrap any instance of TestableTerm that is
+     lambda into `PFun`. Result of this function will be
+     "QuickCheck-compatible" given that all Plutarch type used have
+     `PArbitrary` instance.
 
  @since 2.0.0
 -}
 fromPFun ::
-    forall (a :: S -> Type) (b :: S -> Type) (c :: PTermType).
-    ( c ~ PTT (a :--> b)
-    , FromPFunN a b c
+    forall (p :: S -> Type).
+    ( PUnLam PBool p
+    , PWrapLam (PUnLamHask PBool p)
     ) =>
-    (forall s. Term s (a :--> b)) ->
-    TestableFun c (a :--> b)
-fromPFun = fromPFun' (Proxy @(PTT (a :--> b)))
+    ClosedTerm p ->
+    PLamWrapped (PUnLamHask PBool p)
+fromPFun pf = pwrapLam $ punlam @PBool pf
+
+{- | Ways an Plutarch terms can be compared.
+     @OnPEq@ uses Plutarch `PEq` instance to compare give terms. This
+     means two terms with different UPLC representations can be
+     considered equal when `PEq` instance defines so.
+     @OnUPLC@ uses compiled and evaluated raw UPLC to compare two
+     terms. It is useful comparing Terms that forgot their types--
+     `POpqaue`.
+
+ @since 2.1.0
+-}
+data Equality
+    = OnPEq
+    | OnPData
+    | OnBoth
+    deriving stock (Eq, Show)
```

Comment:
Under what circumstances would `OnBoth` be needed? I'm also unclear what the semantics of `OnBoth` even are.